backendImpl/src/test/java/com/virtuslab/gitmachete/backend/integration/GitMacheteUpstreamInferenceTestSuite.java

```diff
@@ -0,0 +1,71 @@
+package com.virtuslab.gitmachete.backend.integration;
+
+import static org.junit.runners.Parameterized.Parameters;
+
+import lombok.SneakyThrows;
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestWatcher;
+import org.junit.runner.Description;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import com.virtuslab.binding.RuntimeBinding;
+import com.virtuslab.branchlayout.api.IBranchLayout;
+import com.virtuslab.branchlayout.api.manager.IBranchLayoutReader;
+import com.virtuslab.gitmachete.backend.api.IGitMacheteRepository;
+import com.virtuslab.gitmachete.backend.impl.GitMacheteRepositoryFactory;
+import com.virtuslab.gitmachete.testcommon.BaseGitRepositoryBackedTestSuite;
+
+@RunWith(Parameterized.class)
+public class GitMacheteUpstreamInferenceTestSuite extends BaseGitRepositoryBackedTestSuite {
+
+  private final GitMacheteRepositoryFactory gitMacheteRepositoryFactory = new GitMacheteRepositoryFactory();
+  private final IBranchLayoutReader branchLayoutReader = RuntimeBinding
+      .instantiateSoleImplementingClass(IBranchLayoutReader.class);
+  private final IGitMacheteRepository gitMacheteRepository;
+  private final String forBranch;
+  private final String expectedUpstream;
+
+  @Parameters(name = "{0}: inferred upstream of {1} should be {2} (#{index})")
+  public static String[][] getTestData() {
+    return new String[][]{
+        // script name, for branch, expected upstream
+        {SETUP_FOR_OVERRIDDEN_FORK_POINT, "allow-ownership-link", "develop"},
```

Comment:
Not much, but the tests for discover feature will indirectly test upstream inference much more extensively anyway

Reply:
![image](https://user-images.githubusercontent.com/19799111/85985053-63702080-b9ea-11ea-9f8a-3491072258d1.png)